lib/recommendation-server.js

```diff
@@ -0,0 +1,109 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const EXPORTED_SYMBOLS = ['RecommendationServer']; // eslint-disable-line no-unused-vars
+
+function RecommendationServer(opts) {
+  /*
+  Class providing a connection to a recommendation server and a method for
+  querying it. This class should be initialized once and exported as a
+  singleton to modules that will use it.
+
+  Usage:
+
+  let recommendation = new RecommendationServer({
+    // events is an instance of lib/events
+    events: events,
+    xhr: Cc['@mozilla.org/xmlextras/xmlhttprequest;1'],
+    server: 'http://universal-search.dev',
+    timeout: 1000
+  });
+  recommendation.init();
+  */
+  this.events = opts.events;
+  this.server = opts.server;
+  this.timeout = opts.timeout;
+  // note: xhr service, not xhr instance
+  this.xhr = opts.xhr;
+  this.console = opts.console;
+
+  this.search = this.search.bind(this);
```

Comment:
Like I mentioned in the PR description, I have an (admittedly unlikely) dream that this code might live long enough to justify investing in some unit tests. By separating object graph construction / dependency injection from actual initialization, it's really easy to _just_ construct the object in setup methods, and _just_ exercise the initialization code in individual tests, rather than mixing the two.

The current approach is a little awkward, because we're plucking individual deps off the `opts` object, rather than just assigning `this.app`, but the idea there is to be super explicit about dependencies. Also kinda awkward that we're doing `this` binding in the constructor, but Gecko is prone to memory issues, so we really need detachable callbacks, and this is one reasonably palatable way to get there. If we had another month, I'd like to explore Babel and ES6/7 approaches, but we don't.

Does this help? Are you convinced? Let's talk more if not.